lib/vector-core/buffers/src/disk/leveldb_buffer/reader.rs

```diff
@@ -96,18 +99,19 @@ where
             // This will usually complete instantly, but in the case of a large
             // queue (or a fresh launch of the app), this will have to go to
             // disk.
-            let new_data = tokio::task::block_in_place(|| {
-                self.db
-                    .value_iter(ReadOptions::new())
-                    .from(&Key(self.read_offset))
-                    .to(&Key(self.read_offset + 100))
-                    .collect()
+            let mut buffer = std::mem::take(&mut self.buffer);
```

Comment:
I'm confused about why we're doing `mem::take` here. What's going on is we're allocating a fresh, zero-length `VecDeque`, extending the taken buffer by the contents of the value_iter and then placing the taken buffer back into `self.buffer`, trashing the freshly allocated `VecDeque`. Is this to work around a lifetime issue with sending the `buffer` down into `block_in_place`? 

Could we not move the call to `extend` to outside the `block_in_place`?

Reply:
>  Is this to work around a lifetime issue with sending the buffer down into block_in_place?

Exactly. The compiler is having issue with mutable and immutable borrow at the same time from `self` in that function, while creating fresh `VecDeque` is just copying bytes around, there is even a possibility that the compiler is optimizing it away.

> Could we not move the call to extend to outside the block_in_place?

By having `extend` in `block_in_place` we also cover the possibility that the iterator is lazy and actually accessing disk while iterating. I haven't checked for this behavior but either way it seems as a good insurance.

Reply:
Gotcha. At this point it seems like there are a couple of suppositions I'd like to get empirical data for: 

* the compiler is able to optimized the temporary `VecDeque` (I find this believable)
* the iterator is lazy and doing extend internal to `block_in_place` is a performance boon

This function is going to be called enough that I'd like to be sure. We have benchmarks in place for disk buffering specifically, so measuring and approach with and without the temporary would be sufficient to my mind.

Reply:
A, it seams `VecDeque::new` is allocating... yea I'll see what can be changed. At worst I'll extract this change to a separate PR for benchmarking.

Reply:
Cool, here's benchmarks for the disk buffer on my system. I had to rebase with master but otherwise the code here is unchanged. 

![Screenshot_2021-05-28 buffer-disk Summary - Criterion rs](https://user-images.githubusercontent.com/3579/120027277-04bf7800-bfa8-11eb-917e-1c15db030002.png)

Reply:
From these measurements it looks like the median runtimes don't shift all that much for both write-then-read and write-and-read but the distribution is more favorable to the diff version I posted, especially in write-and-read where we'd be hitting the empty condition frequently. You can even see that the modified version I made did fewer overall iterations per benchmark because the results were more stable, which tracks. Removing an allocation from a hot path will give that path more consistent runtime, generally speaking.

Reply:
> The compiler is having issue with mutable and immutable borrow at the same time from self

I've digged for a bit and found it's caused by `Pin`. That's fixed now so this works as expected, without intermediates.